cub/agent/agent_batch_memcpy.cuh

```diff
@@ -0,0 +1,967 @@
+
+    /// The number of bits that each backing data type can store
+    NUM_BITS_PER_UNIT = sizeof(BackingUnitT) * 8,
+
+    /// The number of items that each backing data type can store
+    ITEMS_PER_UNIT = NUM_BITS_PER_UNIT / BITS_PER_ITEM,
+
+    /// The number of bits the backing data type is actually making use of
+    USED_BITS_PER_UNIT = ITEMS_PER_UNIT * BITS_PER_ITEM,
+
+    /// The number of backing data types required to store the given number of items
+    NUM_TOTAL_UNITS = CUB_QUOTIENT_CEILING(NUM_ITEMS, ITEMS_PER_UNIT),
+
+    /// This is the net number of bit-storage provided by each unit (remainder bits are unused)
+    UNIT_MASK = (USED_BITS_PER_UNIT >= (8U * sizeof(uint32_t))) ? 0xFFFFFFFF : (0x01U << USED_BITS_PER_UNIT) - 1,
+    /// This is the bit-mask for each item
+    ITEM_MASK = (BITS_PER_ITEM >= (8U * sizeof(uint32_t))) ? 0xFFFFFFFF : (0x01U << BITS_PER_ITEM) - 1
+  };
+
+  //------------------------------------------------------------------------------
+  // ACCESSORS
+  //------------------------------------------------------------------------------
+public:
+  __device__ __forceinline__ uint32_t Get(uint32_t index) const
+  {
+    int32_t target_offset = index * BITS_PER_ITEM;
+    uint32_t val          = 0;
+
+#pragma unroll
+    for (int32_t i = 0; i < NUM_TOTAL_UNITS; ++i)
+    {
+      // In case the bit-offset of the counter at <index> is larger than the bit range of the current unit,
+      // the bit_shift amount will be larger than the bits provided by this unit.
+      // If the bit-offset of the counter at <index> is smaller than the bit range of the current unit,
+      // the computed bit-offset will be negative, which, once casted to an unsigned type will be larger than the bits
+      // provided by this unit.
+      uint32_t bit_shift = static_cast<uint32_t>(target_offset - i * USED_BITS_PER_UNIT);
+      val |= (data[i] >> bit_shift) & ITEM_MASK;
+    }
+    return val;
+  }
+
+  __device__ __forceinline__ void Add(uint32_t index, uint32_t value)
+  {
+    int32_t target_offset = index * BITS_PER_ITEM;
+
+#pragma unroll
+    for (int32_t i = 0; i < NUM_TOTAL_UNITS; ++i)
+    {
+      // In case the bit-offset of the counter at <index> is larger than the bit range of the current unit,
+      // the bit_shift amount will be larger than the bits provided by this unit.
+      // If the bit-offset of the counter at <index> is smaller than the bit range of the current unit,
+      // the computed bit-offset will be negative, which, once casted to an unsigned type will be larger than the bits
+      // provided by this unit.
+      uint32_t bit_shift = static_cast<uint32_t>(target_offset - i * USED_BITS_PER_UNIT);
+      data[i] += (value << bit_shift) & UNIT_MASK;
+    }
+  }
+
+  __device__ __forceinline__ BitPackedCounter operator+(const BitPackedCounter &rhs) const
+  {
+    BitPackedCounter result;
+    for (int32_t i = 0; i < NUM_TOTAL_UNITS; ++i)
```

Comment:
Why is this one here missing a `#pragma unroll` and should we move to `uint32_t` as the loop index because of conversion warnings against the `uint32_t NUM_TOTAL_UNITS`

Reply:
Thanks! Added the `#pragma unroll` and switched to `uint32_t`. Did you somehow get the compiler to warn about this or did you detect it "manually"? I was not able to get a warning on this with `-Wsign-compare`.